### PR TITLE
Add Focal Loss for chip classification and semantic segmentation

### LINKED ISSUE #979: Implement FocalLoss for Chip Classification
A reference can be found [here](https://arxiv.org/abs/1708.02002).

h/t @AdeelH

### PR DESCRIPTION
**Pull request by AdeelH**
## Overview

Add Focal Loss, as described in the RetinaNet paper, https://arxiv.org/abs/1708.02002. It is essentially an enhancement to cross-entropy loss and is useful for classification tasks when there is a large class imbalance. It has the effect of underweighting easy examples.

- Adds `class FocalLoss(nn.Module)`, which behaves very much like `nn.CrossEntropyLoss()` i.e.
    - supports the `reduction` and `ignore_index` params, and
    - is able to work with 2D inputs of shape `(N, C)` as well as K-dimensional inputs of shape `(N, C, d1, d2, ..., dK)`.
- Example usage
    ```python3
    focal_loss = FocalLoss(alpha, gamma)
	...
	inp, targets = batch
    out = model(inp)
	loss = focal_loss(out, targets)
    ```
- Adds a `focal_loss_gamma` option to `SolveConfig`, which is used as the `γ` for focal loss. The existing `class_loss_weights` option is used as the `α`, if provided.

### Checklist

- [x] Updated `docs/changelog.rst`
- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [ ] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

## Testing Instructions

- How to test this PR
    - Pass e.g. `focal_loss_gamma=2`, with or without specifying `class_loss_weights`, to `SolverConfig` in `spacenet_rio.py` or `isprs_potsdam.py`.

Closes #979 

cc @jamesmcclain

### CODE REVIEW

**Comment by AdeelH** [comment]
- Here's some empirical verification that the two implementations are equivalent for the 2 class case (which the fb implementation is limited to): https://colab.research.google.com/drive/1N9bVs91JyoAMpStECZTanJTrbw3d9JI6?usp=sharing

Some experiment results:
- Spacenet rio CC
    - About the same performance (+0.0009 f1 score) with `alpha = 0.25` and `gamma = 2`: https://s3.console.aws.amazon.com/s3/buckets/raster-vision-ahassan/spacenet_rio/data/focal/
- Potsdam SS
    - Worse performance (-0.002 f1 score) with  no `alpha` and `gamma = 2`: https://s3.console.aws.amazon.com/s3/buckets/raster-vision-ahassan/potsdam/rgb_focal_2/
    - Slightly improvement (+0.0017 f1 score) with  no `alpha` and `gamma = 1`: https://s3.console.aws.amazon.com/s3/buckets/raster-vision-ahassan/potsdam/rgb_focal_1/

**Comment by jamesmcclain** [comment]
> * Here's some empirical verification that the two implementations are equivalent for the 2 class case (which the fb implementation is limited to): https://colab.research.google.com/drive/1N9bVs91JyoAMpStECZTanJTrbw3d9JI6?usp=sharing

Strong.

> Some experiment results:

These numbers look like they could be within the noise.  Do we have any publicly-available datasets with strong imbalances?(Given the comparison to the Facebook implementation above, it looks as though your implementation is equivalent, now I would like to see if this helps our use cases.)

**Comment by AdeelH** [comment]
Superseded by #992
